rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSource.java

```diff
@@ -0,0 +1,324 @@
+        Validate.notEmpty(group, "Consumer group can not be empty");
+
+        if (offsetTable == null) {
+            offsetTable = new ConcurrentHashMap<>();
+        }
+        if (restoredOffsets == null) {
+            restoredOffsets = new ConcurrentHashMap<>();
+        }
+
+        runningChecker = new RunningChecker();
+
+        pullConsumerScheduleService = new MQPullConsumerScheduleService(group);
+        consumer = pullConsumerScheduleService.getDefaultMQPullConsumer();
+
+        consumer.setInstanceName(String.valueOf(getRuntimeContext().getIndexOfThisSubtask()));
+        RocketMqConfig.buildConsumerConfigs(props, consumer);
+    }
+
+    @Override
+    public void run(SourceContext context) throws Exception {
+        LOG.debug("source run....");
+        // The lock that guarantees that record emission and state updates are atomic,
+        // from the view of taking a checkpoint.
+        final Object lock = context.getCheckpointLock();
+
+        int delayWhenMessageNotFound = getInteger(props, RocketMqConfig.CONSUMER_DELAY_WHEN_MESSAGE_NOT_FOUND,
+            RocketMqConfig.DEFAULT_CONSUMER_DELAY_WHEN_MESSAGE_NOT_FOUND);
+
+        String tag = props.getProperty(RocketMqConfig.CONSUMER_TAG, RocketMqConfig.DEFAULT_CONSUMER_TAG);
+
+        int pullPoolSize = getInteger(props, RocketMqConfig.CONSUMER_PULL_POOL_SIZE,
+            RocketMqConfig.DEFAULT_CONSUMER_PULL_POOL_SIZE);
+
+        int pullBatchSize = getInteger(props, RocketMqConfig.CONSUMER_BATCH_SIZE,
+            RocketMqConfig.DEFAULT_CONSUMER_BATCH_SIZE);
+
+        pullConsumerScheduleService.setPullThreadNums(pullPoolSize);
+        pullConsumerScheduleService.registerPullTaskCallback(topic, new PullTaskCallback() {
+
+            @Override
+            public void doPullTask(MessageQueue mq, PullTaskContext pullTaskContext) {
+                try {
+                    long offset = getMessageQueueOffset(mq);
+                    if (offset < 0) {
+                        return;
+                    }
+
+                    PullResult pullResult = consumer.pull(mq, tag, offset, pullBatchSize);
+                    switch (pullResult.getPullStatus()) {
+                        case FOUND:
+                            List<MessageExt> messages = pullResult.getMsgFoundList();
+                            for (MessageExt msg : messages) {
+                                byte[] key = msg.getKeys() != null ? msg.getKeys().getBytes(StandardCharsets.UTF_8) : null;
+                                byte[] value = msg.getBody();
+                                OUT data = schema.deserializeKeyAndValue(key, value);
+
+                                // output and state update are atomic
+                                synchronized (lock) {
+                                    context.collectWithTimestamp(data, msg.getBornTimestamp());
+                                    putMessageQueueOffset(mq, pullResult.getNextBeginOffset());
```

Comment:
Do we need use the next begin offset in other PullStatus?

Reply:
Yes. We need the `nextBeginOffset` in the `PullResult` to update the offset table which will be checkpointing when `snapshotState` invoked .

Reply:
I mean we also should update the offset when the pullresult is OFFSET_ILLEGAL or NO_MATCHED_MSG